File: scripts/drivefolder.py
"""Read a public Drive folder listing with no credentials.

Drive's folder page renders a real HTML table -- one <tr> per child, carrying
the id in data-id and the folder's own ordering in aria-rowindex. No API key,
no OAuth, nothing to rotate or leak.

Folder vs file: a folder reports no size, so Drive labels it "Size not
available" where a file carries "Size: 3.2 MB". That is semantic rather than
cosmetic, unlike the icon class names, which are obfuscated and churn.
"""
import re, html, time, urllib.request, urllib.error

UA = ("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 "
      "(KHTML, like Gecko) Chrome/124.0 Safari/537.36")

class NotShared(Exception):
    """Folder is not readable without signing in (404/403 to an anonymous GET)."""

def fetch(fid, timeout=30, tries=3):
    url = f"https://drive.google.com/drive/folders/{fid}?hl=en"
    req = urllib.request.Request(url, headers={"User-Agent": UA,
                                               "Accept-Language": "en-US,en;q=0.9"})
    last = None
    for n in range(tries):
        try:
            with urllib.request.urlopen(req, timeout=timeout) as r:
                return r.read().decode("utf-8", "replace")
        except urllib.error.HTTPError as e:
            if e.code in (403, 404):
                raise NotShared(f"{fid}: HTTP {e.code} -- folder is not shared publicly")
            last = e
        except Exception as e:
            last = e
        time.sleep(1.5 * (n + 1))
    raise last

def parse(page):
    items = []
    for m in re.finditer(r'<tr [^>]*data-id="([^"]+)"[^>]*aria-rowindex="(\d+)"[^>]*>', page):
        fid, idx = m.group(1), int(m.group(2))
        end = page.find("</tr>", m.end())
        body = page[m.end():end]
        has_size = "Size: " in body
        txt = html.unescape(re.sub(r"\s+", " ", re.sub(r"<[^>]+>", " ", body))).strip()
        name = re.sub(r"^(Video|Image|Audio|PDF|File)\s+", "", txt)
        name = re.sub(r"\s+(Shared|Owned by me|Download).*$", "", name).strip()
        items.append({"i": idx, "id": fid, "name": name,
                      "kind": "file" if has_size else "folder"})
    items.sort(key=lambda x: x["i"])
    return items

def read(fid):
    return parse(fetch(fid))

def read_tree(fid):
    """One level down: files stay files, subfolders come back with their slides.
    That is exactly the folder rule -- a FILE is a reel, a SUBFOLDER is a carousel."""
    out = []
    for it in read(fid):
        if it["kind"] == "folder":
            it = dict(it, slides=[c for c in read(it["id"]) if c["kind"] == "file"])
        out.append(it)
    return out

if __name__ == "__main__":
    import sys, json
    print(json.dumps(read_tree(sys.argv[1]), indent=1))

File: api/drive-folder.js
/* Lists a public Drive folder so a carousel can be built from a link alone.
 *
 * Why scrape rather than call the Drive API: the folder page renders a real
 * HTML table -- one <tr> per child carrying the id in data-id -- so this needs
 * no API key, no OAuth, nothing to rotate or leak. Same approach as
 * scripts/drivefolder.py, which has held up; this is its serverless twin so
 * the browser can do it live instead of at sync time.
 *
 * Folder vs file is semantic, not cosmetic: a folder reports no size, so Drive
 * writes "Size not available" where a file writes "Size: 3.2 MB". The icon
 * class names are obfuscated and churn, so they are not used.
 *
 * Nothing is stored. Bytes are never proxied here -- that is drive-video.js.
 */
const UA =
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 " +
  "(KHTML, like Gecko) Chrome/124.0 Safari/537.36";

const IMAGE = /\.(jpe?g|png|gif|webp|avif|heic|heif|tiff?|bmp)$/i;
const VIDEO = /\.(mp4|mov|m4v|webm|avi|mkv|mpe?g|3gp)$/i;

function parse(page) {
  const items = [];
  const re = /<tr [^>]*data-id="([^"]+)"[^>]*aria-rowindex="(\d+)"[^>]*>/g;
  let m;
  while ((m = re.exec(page))) {
    const id = m[1];
    const end = page.indexOf("</tr>", re.lastIndex);
    const body = page.slice(re.lastIndex, end === -1 ? undefined : end);
    const isFile = body.includes("Size: ");

    let name = body
      .replace(/<[^>]+>/g, " ")
      .replace(/\s+/g, " ")
      .replace(/&amp;/g, "&").replace(/&lt;/g, "<").replace(/&gt;/g, ">")
      .replace(/&quot;/g, '"').replace(/&#39;/g, "'")
      .trim()
      .replace(/^(Video|Image|Audio|PDF|File)\s+/, "")
      .replace(/\s+(Shared|Owned by me|Download).*$/, "")
      .trim();

    const kind = IMAGE.test(name) ? "image" : VIDEO.test(name) ? "video" : null;
    if (isFile && kind) items.push({ id, name, kind });
  }
  return items;
}

export default async function handler(req, res) {
  const id = req.query.id;
  if (!id || !/^[\w-]{10,}$/.test(id)) {
    res.status(400).json({ error: "missing or invalid folder id" });
    return;
  }

  let page;
  try {
    const r = await fetch(`https://drive.google.com/drive/folders/${id}?hl=en`, {
      headers: { "User-Agent": UA, "Accept-Language": "en-US,en;q=0.9" },
      redirect: "follow",
    });
    if (r.status === 403 || r.status === 404) {
      res.status(404).json({ error: "not-shared" });
      return;
    }
    if (!r.ok) {
      res.status(502).json({ error: `drive returned ${r.status}` });
      return;
    }
    page = await r.text();
  } catch (e) {
    res.status(502).json({ error: "could not reach drive" });
    return;
  }

  const files = parse(page);
  /* Ascending by file name, numeric-aware, so IMG_2 lands before IMG_10. */
  files.sort((a, b) => a.name.localeCompare(b.name, undefined, { numeric: true, sensitivity: "base" }));

  /* Folder contents change rarely; a short shared cache keeps repeat reviews
     off Drive's rate limit without making a rename take a day to show up. */
  res.setHeader("Cache-Control", "public, s-maxage=300, stale-while-revalidate=3600");
  res.status(200).json({ folder: id, count: files.length, files });
}
